script/model/forecast.py

```diff
@@ -0,0 +1,585 @@
+                    seq = self.x_transformer.transform(seq.reshape(-1, 1))
+                    labels = self.y_transformer.transform(
+                        labels.reshape(1, -1))
+
+                seq = torch.FloatTensor(seq).view(-1)
+                labels = torch.FloatTensor(labels).view(-1)
+
+                y_pred = self(seq)
+
+                single_loss = loss_function(y_pred, labels)
+                single_loss.backward()
+                optimizer.step()
+
+            if i % 25 == 0:
+                LOG.info(
+                    f'[LSTM FIT]epoch: {i+1:3} loss: {single_loss.item():10.8f}')
+
+        LOG.info(
+            f'[LSTM FIT]epoch: {epochs:3} loss: {single_loss.item():10.10f}')
+
+    def test(self, seq: np.ndarray) -> float:
+        """
+        Perform inference on a dataset. Returns a list of prediction results
+        :param seq: Sequence for testing
+        :return: Prediction results
+        """
+        # To tensor
+        if self.x_transformer:
+            seq = self.x_transformer.transform(seq.reshape(-1, 1))
+        seq = torch.FloatTensor(seq).view(-1)
+
+        with torch.no_grad():
+            model.hidden = (torch.zeros(1, 1, model.hidden_layer_size),
+                            torch.zeros(1, 1, model.hidden_layer_size))
+            pred = model(seq)
+
+        if self.y_transformer:
+            pred = self.y_transformer.inverse_transform(pred.reshape(1, -1))
+
+        return pred.item()
+
+
+if __name__ == "__main__":
+    args = argp.parse_args()
+
+    if args.test_file is None:
+
+        # Generate OLTP trace file
+        if args.gen_data:
+            gen_oltp_trace(
+                tpcc_weight=args.tpcc_weight,
+                tpcc_rates=args.tpcc_rates,
+                pattern_iter=args.pattern_iter)
+
+            trace_file = DEFAULT_QUERY_TRACE_FILE
+        else:
+            trace_file = args.trace_file
+
+        data_loader = DataLoader(
+            query_trace_file=trace_file,
+            interval_us=INTERVAL_MICRO_SEC)
+        x_transformer, y_transformer = data_loader.get_transformers()
+
+        # FIXME: Assuming all the queries in the current trace file are from
+        # the same cluster for now
+        cluster_trace = QueryCluster(data_loader.get_ts_data())
+
+        # Aggregated time-series from the cluster
+        train_timeseries = cluster_trace.get_timeseries()
+
+        # Perform training on the trace file
+        lstm_model = LSTM(
+            lr=args.lr,
+            epochs=args.epochs,
+            x_transformer=x_transformer,
+            y_transformer=y_transformer)
+        trainer = ForecastTrainer(
+            train_timeseries,
+            seq_len=args.seq_len,
+            eval_size=args.eval_size,
+            horizon_len=args.horizon_len)
+        trainer.train([lstm_model])
+
+        # Save the model
+        if args.model_save_path:
+            with open(args.model_save_path, "wb") as f:
+                pickle.dump(lstm_model, f)
```

Comment:
Move these logics inside `Forecaster`, e.g., inside the `train` method.

Reply:
Will this force a forecaster to be coupled with all clusters? 
I was leaving all this logic here 1) for just quick prototyping since I think this will change once we have multiple clusters. 2) a forecaster should be corresponding to one cluster, but a trace file might contain multiple clusters. So clusters' data should be injected into a `Forecaster` 

Or the `Forecaster` would be dealing with multiple clusters internally, so it will have multiple sets of timeseries and models (one for a cluster) for multiple clusters.

Reply:
This is more like a conceptual thing. But yes, I was thinking of the second architecture in mind. The `Forecaster` would deal with multiples clusters (and perform the clustering in the future) internally.

Reply:
Cool, I could expand on this architecture.